[ShadowElevations] Update ShadowElevations to be extensible.

mpetrov opened the pull request:
This defines a custom macro to get the typed enum behavior rather than defining a system macro to prevent it from adversely affecting clients.
Redundant `NS_SWIFT_NAME` macros are removed. 
Usage is updated to omit the optional `rawValue:` argument, though this is not strictly necessary since it still conforms to `RawRepresentable`. Docs are also updated to reflect the change.

See issue #285;

mpetrov committed:
[Collections] Modernize Swift in styling readme.

mpetrov committed:
Merge remote-tracking branch 'upstream/develop' into develop

mpetrov committed:
[ShadowElevations] Update ShadowElevations to be extensible.

This defines a custom macro to get the typed enum behavior rather than
defining a system macro to prevent it from adversily affecting clients.

Redundant NS_SWIFT_NAME macros are removed.

Usage is updated to omit the optional rawValue: argument. Docs are also
updated to reflect the change.

mpetrov committed:
minor fixes

willlarche reviewed (approved):
Thank you!!!

jverkoey commented:
This doesn't build in Xcode 8.3.3 unfortunately, but it does if I add back the `__attribute__` conditional path:

```swift

#ifdef NS_TYPED_EXTENSIBLE_ENUM // This macro is introduced in Xcode 9.
#define MDC_SHADOW_ELEVATION_TYPED_EXTENSIBLE_ENUM NS_TYPED_EXTENSIBLE_ENUM
#elif defined(CF_TYPED_EXTENSIBLE_ENUM) // What follows is backwards compat for Xcode 8 and below.
#define MDC_SHADOW_ELEVATION_TYPED_EXTENSIBLE_ENUM CF_TYPED_EXTENSIBLE_ENUM
#elif __has_attribute(swift_wrapper)
#define MDC_SHADOW_ELEVATION_TYPED_EXTENSIBLE_ENUM __attribute__((swift_wrapper(struct)))
#else
#define MDC_SHADOW_ELEVATION_TYPED_EXTENSIBLE_ENUM
#endif
```

jverkoey commented on components/ShadowElevations/src/MDCShadowElevations.h:
```diff
@@ -18,90 +18,83 @@
+#ifdef NS_TYPED_EXTENSIBLE_ENUM // This macro is introduced in Xcode 9.
+#define MDC_SHADOW_ELEVATION_TYPED_EXTENSIBLE_ENUM NS_TYPED_EXTENSIBLE_ENUM
+#elif defined(CF_TYPED_EXTENSIBLE_ENUM) // What follows is backwards compat for Xcode 8 and below.
+#define MDC_SHADOW_ELEVATION_TYPED_EXTENSIBLE_ENUM CF_TYPED_EXTENSIBLE_ENUM
```
I think we need another logical path here for Xcode 8.3.3:

```swift
#ifdef NS_TYPED_EXTENSIBLE_ENUM // This macro is introduced in Xcode 9.
#define MDC_SHADOW_ELEVATION_TYPED_EXTENSIBLE_ENUM NS_TYPED_EXTENSIBLE_ENUM
#elif defined(CF_TYPED_EXTENSIBLE_ENUM) // What follows is backwards compat for Xcode 8 and below.
#define MDC_SHADOW_ELEVATION_TYPED_EXTENSIBLE_ENUM CF_TYPED_EXTENSIBLE_ENUM
#elif __has_attribute(swift_wrapper)
#define MDC_SHADOW_ELEVATION_TYPED_EXTENSIBLE_ENUM __attribute__((swift_wrapper(struct)))
#else
#define MDC_SHADOW_ELEVATION_TYPED_EXTENSIBLE_ENUM
#endif
```

mpetrov committed:
need an extra clause for xcode8

mpetrov commented:
@jverkoey - thanks, you're right about the 8.3.3 case. I could have sworn I tried that, but maybe I built the wrong target or had some statefulness at play.
I could also completely nuke the `CF_TYPED_EXTENSIBLE_ENUM` clause, since the NS_ seems preferred in Apple's documentation.

ianegordon reviewed (approved):
Travis xcode 8 builds are happy again.
LGTM.
I'll leave it to Jeff in case he has any last requests.

jverkoey commented on components/ShadowElevations/src/MDCShadowElevations.h:
```diff
@@ -18,90 +18,85 @@
-#define CF_TYPED_ENUM __attribute__((swift_wrapper(enum)))
+#ifdef NS_TYPED_EXTENSIBLE_ENUM // This macro is introduced in Xcode 9.
+#define MDC_SHADOW_ELEVATION_TYPED_EXTENSIBLE_ENUM NS_TYPED_EXTENSIBLE_ENUM
+#elif defined(CF_TYPED_EXTENSIBLE_ENUM) // What follows is backwards compat for Xcode 8 and below.
```
Can you verify that these comments are accurate now? I think they need to be updated.

jverkoey reviewed (approved):
LGTM pending comments being cleaned up.

mpetrov committed:
updated for xcode8 backwards compat

mpetrov committed:
merge

mpetrov commented:
Opted for either using `NS_TYPED_EXTENSIBLE_ENUM` or just falling back to the `swift_wrapper(struct)` attribute. I removed the `CF_TYPED_EXTENSIBLE_ENUM` since I couldn't find references to it in Apple's docs, and it doesn't seem to be present in Xcode8 headers.

I removed `Availability.h` (tried `CFAvailability.h` as well) since these didn't seem to change any of the macros used. Instead relying on `NS_TYPED_EXTENSIBLE_ENUM` being defined in Foundation.

jverkoey commented:
Rad - thanks for ironing this out!
